File: TypeScript/modules/supertoaster.ts
import { Toaster } from "./toaster.ts";
import { toastZustandAlsString, delay } from "./util.ts"

export class SuperToaster extends Toaster {

    private temperatur = 0

    constructor(farbe: string, schaechte: number) {
        // Super sorgt dafür das der der Construktor aus der Superklasse ausgeführt wird.
        super(farbe, schaechte);
    }

    // Hier wird toasten() overridden und eine extra Konstrollstruktur hinzugefügt welche die Temperatur prüft.
    async toasten(): Promise<void> {
        if (this.anzahlToast == 0) {
            console.log("Geht nicht. Kein Toast drin.");
        } else if (this.temperatur >= 500) {
            console.log("Breche Vorgang ab. Zu heiß.");
        } else {
            console.log("Toasten...");
            await delay(this.toastZeit * 1000)
            if (this.toastZeit == 0) {
                this.toastZustand = 0
            } else if (this.toastZeit <= 15) {
                this.toastZustand = 1
            } else if (this.toastZeit > 15) {
                this.toastZustand = 2
            } else if (this.toastZeit > 30) {
                this.toastZustand = 3
            }
            // Mir fällt nichts besseres ein um eine "Überhitzung" zu simulieren.
            this.temperatur += (5 * this.toastZeit)
            console.log(`Fertig. Toast ist: ${toastZustandAlsString(this.toastZustand)}`); //toastZustandAlsString() ist eine utility Funktion die ich ausgelagert habe. Mehr dazu siehe util.ts.
            console.log(`Toastertemperatur beträgt: ${this.temperatur}`);
        }
    }
}

File: Python/supertoaster.py
from time import sleep
from toaster import Toaster
from toaster import toast_zustand_als_string

class SuperToaster(Toaster):

    def __init__(self, farbe, schaechte):
        # super() sorgt dafür das die Properties und Methoden von Toaster übernommen werden.
        super().__init__(farbe, schaechte)
        self._temperatur = 0


    # Override der toasten() Methode. Grund: Check der Temperatur zum abbrechen des Toastvorgangs.
    def toasten(self):
        if self._anzahl_toasts == 0:
            print("Geht nicht. Kein Toast drin.")
        elif self._temperatur >= 500:
            print("Breche Vorgang ab. Zu heiß.")
        else:
            print("Toasten...")
            sleep(self._toastzeit)
            if self._toastzeit == 0:
                self._toasts_zustand = 0
            elif self._toastzeit <= 15:
                self._toasts_zustand = 1
            elif self._toastzeit > 15:
                self._toasts_zustand = 2
            elif self._toastzeit > 30:
                self._toasts_zustand = 3
            print(f"Fertig. Toast ist: {toast_zustand_als_string(self._toasts_zustand)}");